File: src/vip_tests/workbench/chronicle_probe.R
# In-session Chronicle raw-chunk probe. Loaded and flattened to a single console
# line by chronicle_probe.py; the test sends it via rstudio_eval and reads back
# the sentinel token this function returns.
#
# Chronicle writes each metric's raw telemetry to
#   <base_path>/<store>/<server-id>/v2/<metric>/<Y>/<M>/<D>/<H>/<chunk>.csv
# where <store> is "private" or "hourly". Chunk filenames are unstable (mostly
# bare UUIDs, sometimes a "chronicle-data-*" stem), so we match any .csv/.parquet
# file under the metric's "/v2/<metric>/" directory rather than a filename prefix.
# The trailing slash stops a metric from matching a longer one that shares its
# prefix. We read the newest non-empty chunk to confirm the session user can
# actually read collected rows. We first require the chunk to be readable at all
# (file.access mode 4) so a chunk the session user cannot open -- e.g. permission
# denied -- reports VIP_NO_DATA rather than a false positive. n is then its line
# count (NA when the readable file is not text, e.g. a Parquet chunk), so a CSV
# with a header plus >= 1 row has n > 1.
#
# Returns exactly one sentinel token:
#   "VIP_DATA_OK"   a non-empty chunk exists and holds >= 1 data row
#   "VIP_NO_DATA"   no readable, non-empty chunk exists for the metric
#
# chronicle_probe.py flattens this file by dropping comment/blank lines and
# joining the rest with "; ", so every line below must be a complete R statement:
# no line-spanning calls, no bare "} else", and no inline "#" comments.
vip_chronicle_probe <- function(base_path, metric) {
  files <- list.files(base_path, pattern = "[.](csv|parquet)$", recursive = TRUE, full.names = TRUE)
  files <- files[grepl(paste0("/v2/", metric, "/"), files, fixed = TRUE)]
  files <- files[file.info(files)$size > 0]
  if (length(files) == 0) return("VIP_NO_DATA")
  newest <- files[order(file.info(files)$mtime, decreasing = TRUE)][[1]]
  if (file.access(newest, mode = 4) != 0) return("VIP_NO_DATA")
  n <- tryCatch(length(readLines(newest, warn = FALSE)), error = function(e) NA_integer_)
  if (is.na(n) || n > 1) "VIP_DATA_OK" else "VIP_NO_DATA"
}

File: src/vip_tests/workbench/chronicle_probe.py
"""Builder for the in-session Chronicle raw-chunk probe expression.

The probe logic lives in ``chronicle_probe.R`` as a normal, readable R function.
``rstudio_eval`` types the expression into the console as a single line, so this
module flattens that file into one line (dropping comment/blank lines and joining
statements with ``"; "``) and appends the parameterized call. Kept separate from
``test_chronicle.py`` (which carries the pytest-bdd ``@scenario`` and Playwright
imports) so the builder can be unit tested without a browser — see
``selftests/test_workbench_chronicle.py``.
"""

from __future__ import annotations

from pathlib import Path

# Sentinel tokens the probe returns (exactly one). Must match the string
# literals in chronicle_probe.R; a selftest guards that they stay in sync.
TOKEN_OK = "VIP_DATA_OK"
TOKEN_NO_DATA = "VIP_NO_DATA"

_PROBE_R_FILE = Path(__file__).with_name("chronicle_probe.R")


def _flatten_r(source: str) -> str:
    """Collapse an R source file to a single console-safe line.

    Drops blank lines and full-line ``#`` comments, then joins the remaining
    statements with ``"; "``. This is safe only because ``chronicle_probe.R`` is
    written with one complete statement per line and no inline comments (see the
    header comment there).
    """
    lines = [line.strip() for line in source.splitlines()]
    statements = [line for line in lines if line and not line.startswith("#")]
    return "; ".join(statements)


# Flatten once at import; the R file is static (the call is parameterized below).
_PROBE_FN = _flatten_r(_PROBE_R_FILE.read_text())


def raw_chunk_probe_expr(base_path: str, metric: str) -> str:
    """Return the one-line R expression that probes one raw Chronicle metric.

    Defines ``vip_chronicle_probe`` (from chronicle_probe.R) and calls it for
    *base_path* / *metric*, printing exactly one sentinel token (``TOKEN_OK`` or
    ``TOKEN_NO_DATA``). ``base_path`` and ``metric`` are embedded as R string
    literals; callers must reject values containing a double quote (see
    ``test_chronicle._safe_r_literal``).
    """
    return f'{_PROBE_FN}; cat(vip_chronicle_probe("{base_path}", "{metric}"))'
